pkg/sentry/fs/tty/line_discipline.go

```diff
@@ -203,6 +203,24 @@ func (l *lineDiscipline) inputQueueRead(ctx context.Context, dst usermem.IOSeque
 	return 0, syserror.ErrWouldBlock
 }
 
+func (l *lineDiscipline) discardPendingInput(ctx context.Context) (int64, error) {
+	l.termiosMu.RLock()
+	defer l.termiosMu.RUnlock()
+	var total int64 = 0
+	outBytes := make([]byte, 32)
+	for {
+		dst := usermem.BytesIOSequence(outBytes)
+		n, err := l.inputQueueRead(ctx, dst)
+		total += n
+		if err != nil {
+			if err == syserror.ErrWouldBlock {
+				return total, nil
+			}
+			return total, err
+		}
+	}
+}
```

Comment:
Two things about this:

1. There's a race condition. After the function returns, but before `setTermios` is called, bytes can be written to the input queue and read out the other side. This could manifest as racy behavior where some bytes are discarded, but bytes written later are delivered with the old termios transformations applied.
2. It looks like this is repeatedly copying bytes just to discard. This seems like an unnecessarily expensive way of discarding input.

To deal with both issues, I wonder whether `TCSETSF` could lead to a call to a new function (`setTermiosFlush` for example) that atomically discards input and clears the appropriate queue. It would likely look something like:

```go
func (l *lineDiscipline) setTermiosFlush(/*args*/) (uintptr, error) {
	l.termiosMu.Lock()
	defer l.termiosMu.Unlock()
	// Clear the queue directly...
	setTermiosLocked()
}

func (l *lineDiscipline) setTermios(/*args*/) (uintptr, error) {
	l.termiosMu.Lock()
	defer l.termiosMu.Unlock()
	setTermiosLocked()
}

func (l *lineDiscipline) setTermiosLocked(/*args*/) (uintptr, error) {
	// The old setTermios implementation, minus locking.
}
```

Reply:
Good idea, i've added something along this line.